### ISSUE TITLE
Desc not exposed in GaugeVec

### ISSUE THREAD

**Issue by lledey** [prometheus/client_golang#847]
**TL;DR**: GaugeVec doesn’t expose its `*Desc`, I can’t do error handling with `NewInvalidMetric` in my custom collector

Hi!

I’m running into some issues when trying to use a GaugeVec and a custom Collector.

Some of the metrics I’m sending are heavy to compute, so I implemented a custom collector to be able to control when the metrics are sent (every hour, every day,…) instead of sending them every time the `Collect` method is called.
This means that in the `Collect` method in the custom Collector, I’m first checking if the threshold to update the metrics is reached, and if that’s the case then I send my metrics and collect them.
When sending the metrics, some errors can happen when computing the gauge values for example, so I have some error handling, and it worked perfectly since I’ve been using Gauge only. It gives something like that
```go
if err := metric.ComputeAndSet(); err != nil {
	switch errors.Cause(err) {
	// some cases
	default:
		ch <- prometheus.NewInvalidMetric(metric.Desc(), err)
		return
	}
}
```

My issue here is when I try to use a GaugeVec, because even though internally it uses a single `*Desc`, it is not exposed, meaning that I can’t call NewInvalidMetric in case of error.

Maybe I’m missing something, but I think the API is not consistent between Gauge and GaugeVec, and I think the `*Desc` should be exposed by GaugeVec.

I can try to submit a PR to expose the `*Desc` in GaugeVec if needed, but I’m not sure if `*metricsMap` should be exposed or if caching the `*Desc` in the GaugeVec instance directly and adding a `Desc()` method would work (or maybe another solution I didn't think of).

Thanks!

**Comment by beorn7** [comment]
`GaugeVec` has a `Describe(chan<- *Desc)` method. It's a bit clunky for your use case, but you can use it to retrieve the Desc.

**Comment by lledey** [comment]
True, didn't think about that! I'm not sure if `Collect` can be called before `Describe` though, but it feels very hacky anyway.

I believe the true fix would be to expose a `Desc` method like in `Gauge` and the other types. How would you feel about that instead?

**Comment by beorn7** [comment]
The `Desc` method is part of the `Metric` interface, i.e. a single metric child, producing a single metric with the `Write` call and therefore returning exactly one `Desc` via the `Desc` method. All the various XxxVec types are implementing `Collector`, though. They return a variable number of `Metric`s and therefore _may_ return more than one `Desc` (or even zero, FWIW).

Special-casing a `Collector` that happens to return only one `Desc`, just to cater for a quite unusual use case, is a bit clunky. Most importantly, it burdens all other users of the library with a proliferation of methods, of which some do the same as others, just with a different way of returning the results.

(In different news, the fact that `Describe` uses a channel to return the results is a result of naive channel euphoria from the early days of Go. Your use case would have been easier to deal with if it just returned a `[]*Desc`, which the next major release will do anyway. See #228 .)

**Comment by lledey** [comment]
I don't get it.

Implementing and using a custom collector may not be the most usual use-case, but I can't believe it's an edge-case either. And even though it was an edge-case, half of the lib is unusable here, which is an issue.

Looking at the implementation of every XxxVec collectors (`GaugeVec`, `CounterVec`, `SummaryVec`, `HistogramVec`), they all create a new `Desc` and share it in the underlying `MetricVec`. It's even specified in [the doc](https://pkg.go.dev/github.com/prometheus/client_golang/prometheus#GaugeVec) of all of these types that `XxxVec is a Collector that bundles a set of Xxx that all share the same Desc`.
All these 4 types could definitely expose their `Desc` as per the doc, so that, again, their API would be consistent with the rest of the lib, and they would be useable by custom collectors. I don't see how this would be a burden for users and be interpreted as a proliferation of methods, especially since the prototype would be exactly the same everywhere. If users don't need the method, they just won't use it. And for use-cases like mine, I need it and I'll use it, and everybody is happy.

I'm not doing this for fun, we're all doing business here, and right now I'm stuck because of this. I can't wait for a v2 either that is being discussed since 2016.
I get it if you don't have the time to implement the change, we're all busy, but I'm volunteering to submit a PR. Why the push-back?

**Comment by beorn7** [comment]
All you need is a three-liner:

```Go
func GetDesc(v *prometheus.MetricVec) *prometheus.Desc {
	ch := make(chan *prometheus.Desc, 1)
	v.Describe(ch)
	return <-ch
}
```

I wouldn't call that "half of the lib is unusable here".

I also didn't call custom collectors unusual, but requiring to extract the `Desc` from one of the pre-defined vector types is unusual.

More generally, we seem to have different opinions of consistency and what's a burden for the users. That's fine, people can have different opinions. I tried to explain mine as well as I could. Even if you don't share my opinion or "don't get it", how about taking a leap of faith that, after all those years, I have thought about these things a lot and have processed a lot of user feedback? Fatalistic language ("half of the lib is unusable here"), and especially your last paragraph isn't helpful at all in this context.

**Comment by lledey** [comment]
Thanks for the answer. I realize my tone wasn't nice, I apologize for that. I guess part of it was a misunderstanding and some frustration that after 2 back and forth I felt things weren't moving forward. Again sorry about that.

I was gonna write this message even without your answer, but thanks for the help and the snippet, we'll probably make things work this way 🙏
